P2 integration

kcammel opened the pull request:
Changes for P2 integration of live sturgeon processing.

- Extra software requirement: modkit needs to be installed on pc
- new venv compared to current sturgeon

Main differences with gridion live sturgeon:

- Basecalling and alignment is performed by the sequences, this application only performs the sturgeon commands and extra plotting for the java gui
- R_scripts are mainly replaced with python_scripts, for more practicality and flexibility with talking to system paths etc\
- Java gui is also changed (https://gitlab.researchidt.prinsesmaximacentrum.nl/ajanse/sturgeon-gui/-/tree/P2_integration?ref_type=heads)
- Lock file and running process is streamlined (wrt wrapper_stop.txt)

For testing purposes:
I created the file dummy_output_generator.py (in testing_purposes dir). This creates a symlink to a bam file in a specified path (https://github.com/princessmaximacenter/sturgeon/blob/P2_integration/testing_purposes/dummy_output_generator.py#L7), in a output directory bam_pass (make this directory be the location for the "pod5 input folder" in the java gui. First symlink is created after 20 seconds, then with 2 minute intervals. Files are from the prom@P22 /data/20250417_CNS_test_p3

kcammel committed:
Python scripts to watch for bam output files by the P2 and process with sturgeon

kcammel committed:
Scripts for live processing P2 results (bam files)

kcammel committed:
Updated readme for required installation of modkit or Rust/Cargo

kcammel committed:
Updated hard-coded paths to R scripts to be imported through config file

kcammel committed:
Script to simulate creation of bam files in output dir, for testing purposes

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+DEFAULT_SUFFIXES: str = ".bam"
+DEFAULT_LOCK_FILE: Path = Path("script.lock")
+DEFAULT_MODEL_FILE: Path = Path("/Users/k.v.cammel/Developer/cold_setup_sturgeon/sturgeon/include/models/general.zip")
+DEFAULT_CONFIG_FILE: Path = Path("/Users/k.v.cammel/Developer/cold_setup_sturgeon/python_scripts/config.yaml")
```
I would stored all these values in the config file and make a config class to retrieve these values

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+# ==========================
+# File Event Handler
+# ==========================
+class NewBamFileHandler(FileSystemEventHandler):
```
It is better if a class has their own dedicated file

AlexJanse replied:
Also good to think about which functions and variables need to be public and which needs to be private. In python it is not a hard rule but it may help developers later to understand your code better. You make something private by adding a lower dash in the variable/function name i.e. `self._private=1`. A good python editor would warn the developer if they are accessing something they shouldn't have.

AlexJanse replied:
Another thing to consider is which functions can be static (a.k.a. which function do not need the self object).
You can add `@staticmethod` above the def statement to mark it as such. This also just a best practice thing to do. And you can access these method without making an object out of the class (Most useful for my comment in the plotting script).

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+    """Check and create a lock file to prevent multiple instances."""
+    if lock_file.exists():
+        print("A different instance of Sturgeon is already running. Exiting...")
+        sys.exit(1)
```
Have you checked if this error is raised if the GUI also acts as it should? Perhaps it is also better to raise an error instead of printing it. If you print something it might end up in the standard output instead of the standard error log. An example on how to so could be:

```python
import logging  # this is the package we usually use for making log statements instead of print
log = logging.getLogger('root')

log.info("Checking availability sturgeon")

# style 1
log.error("A different instance ....")
sys.exit(1)

#style 2
raise LookupError("A different instance ....")
```
Think using log instead of print throughout your code might be the best way to go as you are certain where it ends up

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+
+# ==========================
+# CLI with Click
+# ==========================
```
I have never heard of click before but looks cool 😁 If possible I would also put these inside a function but given the syntax not sure if possible

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+        self.r_script_path = self.config['paths'].get('r_script_main')
+
+
+    def load_config(self, config_yaml):
```
missing input and output type

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+            sys.exit("Exiting...")
+
+    def on_created(self, event: FileSystemEvent) -> None:
+        """React to newly created files that match allowed suffixes."""
```
Although it might be overkill but still a good Python practice: adding docString. This explains what each input is about as well as the function. Pycharm can make you a template by changing the following settings:
pycharm (menu bar) -> settings -> Tools -> Python Integrated Tools -> Docstring format -> reStructuralText

pycharm (menu bar) -> settings -> Editor -> general -> Smart Keys -> Insert documentation comment stub (make sure it is checked)

Then if you type triple double qoutes and hit enter, it should make you a template that you can fill in. It might not be needed for every function but at least use it to public functions

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+import sys
```
My advise for your python code in general is to decide:

- style: either use OOP style (so make sure almost everything is within classes; also preferred style within the group) or don't (so everything inside functions but not in classes or global). Although possible, global variables are a big no go in python best practices so unless absolutely necessary we would like to avoid those.
- Separate logic from front end. It is better to have a script that handles the user input and calls the logic in another script rather than having everything inside one file. This way others can use your logic directly without being forced into your front end implementation. This is very optional given the file is quite small but good to keep it in mind in the future.

AlexJanse commented on python_scripts/live_processing_P2I.py:
```diff
@@ -0,0 +1,212 @@
+        output_file = f"{self.output}/iteration_{self.iteration}/confidence_over_time_plot_iteration_{self.iteration}"
+        df = pd.read_csv("/Users/k.v.cammel/Documents/Projects/P2_wrapper/color_translation.csv")
+        color_translation = dict(zip(df["class"], df["color"]))
+        plot_confidence_over_time(self.full_data, output_file, color_translation)
```
This is a personal preference so feel free to ignore: for readability it might be nice to have some indication where something is coming from. so instead of
```python
from live_sturgeon_plotting import plot_confidence_over_time, write_progress_tsv, plot_CNV_bam
# ....
plot_confidence_over_time()
```
you can decide to do:
```python
import live_sturgeon_plotting as lsp
# ....
lsp.plot_confidence_over_time()
```
The reason why I prefer this is because it makes it clear that this function is not a class function but a function of somewhere else without me having to look for the declaration/import of it.

AlexJanse commented on python_scripts/live_sturgeon_plotting.py:
```diff
@@ -0,0 +1,101 @@
+from unittest.mock import inplace
```
General comment: if you go for OOP style, you can put these functions inside a class and make the functions static (as mentioned in my general comment of the P2I script)

AlexJanse commented on utils/CNV_plot.R:
```diff
@@ -0,0 +1,108 @@
+suppressMessages(library(DNAcopy))
+if (!requireNamespace("BiocManager", quietly = TRUE))
+    install.packages("BiocManager")
+suppressMessages(BiocManager::install("QDNAseq"))
```
A rule in R is that you don't install packages for the users. The user either need to install it themselves or it is a dependency for a R package which is installed when building the package. You don't have to chance this now but for the future keep this in mind

kcammel replied:
As discusses during code review, for testing and getting V0.1 running on the P2 we will keep this with an addition of update = FALSE to ensure sturgeon can be run. But yes, need to change this for V1.0 release

AlexJanse commented on utils/CNV_plot.R:
```diff
@@ -0,0 +1,108 @@
+  sample <- sample[sample$chrom != "Y" & sample$chrom != "X", ]
+  sample$reltoref <- sample$coverage / sample$refcov
+  blacklistbins <- sample[sample$refcov < 500, ]
+  blacklistbins <- row.names(blacklistbins)
```
Maybe good to change things like blacklist to exclusionlist and whitelist to inclusionlist or something similar.

AlexJanse commented on pyproject.toml:
```diff
@@ -13,17 +13,62 @@ authors = [
-    "onnxruntime==1.12.1",
-    "pysam==0.19.0",
-    "modbampy==0.6.3",
-    "pod5==0.2.4",
```
I see that some versions have changed? Does this affect sturgeon as well?

AlexJanse commented on sturgeon_P2.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
Can you add a brief description on what this file does and how it is used?

AlexJanse commented on testing_purposes/dummy_output_generator.py:
```diff
@@ -0,0 +1,54 @@
+import os
```
Very nice to have this 😲 For the final version of this code make sure your path can be given as input instead of being hard coded

kcammel committed:
Changed variable name and set updating of packages to false to ensure live processing can continue

kcammel committed:
Extra explanation and dealing with installation:
R script:
Do not install R packages each time, but just load path. Correct path on the prom to be added
Bash script:
Added comment

kcammel committed:
Split up python script to handle separately:
-logging
-CLI, exit, turning on main script
-Bam file handling
-Plotting results

Default parameters taken from config.yaml

kcammel committed:
Put modkit results of all iterations in 1 dir. This ensures that sturgeoninputtobed merges all results in that dir, so you get more analyzed probes per new iteration

kcammel committed:
Changed watchdog observer to first check if bam files are already present in results folder and process those.
Added queue to ensure that sturgeon is not overloaded by new results, so bam files can be written every 1 minute

kcammel committed:
Changed it so that when starting gui, technician can put as input folder the sequencing run, and input the relevant barcode. Script will then handle to check for the actual dir with the relevant bam files

kcammel committed:
Add in FLAG to trigger gui waiting box
Add in normalize barcode function so barcode can be given either as (E.g.) 5 or 05

kcammel committed:
Corrected log message for correct button updating in gui

kcammel committed:
Put in option if run is a gridion run, changes some parameters.
Also changed some options to make entering results directory even more foolproof

kcammel committed:
Small bug fix for setting correct results directory

kcammel committed:
Some changes for:
-handling a gridion verification run
-Merging all previous bams for generating CNV plot

kcammel committed:
Small change to only label the most recent condifence value, for better readability

kcammel committed:
Change that merges bam files pre-modkit filtering (to make it more in-line with gridion results)
Make symlinks, to save on copying time

kcammel committed:
Change to make the confidence plot x-axis more readable

AlexJanse commented on python_scripts/SturgeonBamHandling.py:
```diff
@@ -0,0 +1,201 @@
+
+
+    def load_color_translation(self):
+        df = pd.read_csv("/Users/k.v.cammel/Developer/cold_setup_sturgeon/utils/color_translation.csv") #Set in config.yaml
```
hard coded path?

AlexJanse commented on python_scripts/SturgeonLivePlotting.py:
```diff
@@ -0,0 +1,105 @@
+        mgd.rename(columns={"score": f"iteration_{iteration}"}, inplace=True)
+        return mgd
+
+def plot_confidence_over_time(full_data,output_file,color_translation):
```
Code looks great! The final suggestion is to be consistent in parameter description in function definitions. So which type do you expect and what is it. This is not perse needed for private functions but given most of your functions are public it is a nice thing to do. So in this case I expect something like:
```
def plot_confidence_over_time(full_data: pd.DataFrame, output_file: str,  color_translation: dict) -> None:
"""
Write plot confidence over time
y-axis: confidence
x-axis: iteration

:param full_data: dataframe containing ....
:param output_file: Path to output file
:param color_translation: Dict containing per label a color hex code
:return: None 
"""
```

AlexJanse commented on python_scripts/SturgeonLivePrediction.py:
```diff
@@ -0,0 +1,175 @@
+import SturgeonLogging as SL
+
+# Load config file with default values
+CONFIG_PATH = Path("/Users/k.v.cammel/Documents/Projects/P2_wrapper/sturgeon/python_scripts/config.yaml")
```
hard coded path

AlexJanse commented on python_scripts/config.yaml:
```diff
@@ -0,0 +1,13 @@
+# Sturgeon project configuration
+paths:
+  r_script: "/Users/k.v.cammel/Developer/cold_setup_sturgeon/utils/CNV_plot.R"  # Eventually change path when installed on P2
```
personal paths in config

kcammel committed:
Removed hardcoded/personal paths.
Added descriptions to functions

kcammel committed:
Removed some more personal paths
Made dummy_output_generator.py a bit nicer to use

kcammel committed:
Path objects don't always work well with rpy2, so change to str

kcammel committed:
Messed up settings for watchdog:
Explicitily looks for "on_created" so "_on_created" was not valid

Fixed typo in function wait_for_directory

kcammel committed:
Merge remote-tracking branch 'origin/dev' into P2_integration
